bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +87,80 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
+                            ref::compareAndSet).run();
+                }).whenComplete((metadata, ex) -> {
+                        LOG.info("Update of {} finished", ledgerId);
+                        if (ex != null
+                            && !(ex instanceof BKException.BKNoSuchLedgerExistsException)) {
+                            LOG.error("Updating ledger metadata {} failed", ledgerId, ex);
+                            errorOccurred.set(true);
+                        } else {
+                            updatedLedgerCnt.incrementAndGet();
+                            progressable.progress(updatedLedgerCnt.get(), issuedLedgerCnt.get());
                         }
-                        bkc.getLedgerManager().readLedgerMetadata(lId, readCb);
-                    }
-                    // waiting till all the issued ledgers are finished
-                    syncObj.await();
-                } catch (IOException ioe) {
-                    LOG.error("Exception while updating ledger", ioe);
-                    throw new RuntimeException("Exception while updating ledger", ioe.getCause());
-                } catch (InterruptedException ie) {
-                    LOG.error("Exception while updating ledger metadata", ie);
-                    Thread.currentThread().interrupt();
-                    throw new RuntimeException("Exception while updating ledger", ie.getCause());
-                }
-            }
-        });
-        try {
-            // Wait to finish the issued ledgers.
-            updateBookieCb.get();
-        } catch (ExecutionException ee) {
-            throw new IOException("Exception while updating ledger", ee);
-        } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt();
-            throw new IOException("Exception while updating ledger", ie);
-        } finally {
-            executor.shutdown();
-        }
-    }
+                    });
 
-    private static final class UpdateLedgerCb implements FutureCallback<Void> {
-        final long ledgerId;
-        final AtomicBoolean stop;
-        final AtomicInteger issuedLedgerCnt;
-        final AtomicInteger updatedLedgerCnt;
-        final Set<Long> outstandings;
-        final CountDownLatch syncObj;
-        final UpdateLedgerNotifier progressable;
-
-        public UpdateLedgerCb(long ledgerId, AtomicBoolean stop, AtomicInteger issuedLedgerCnt,
-                AtomicInteger updatedLedgerCnt, Set<Long> outstandings, CountDownLatch syncObj,
-                UpdateLedgerNotifier progressable) {
-            this.ledgerId = ledgerId;
-            this.stop = stop;
-            this.issuedLedgerCnt = issuedLedgerCnt;
-            this.updatedLedgerCnt = updatedLedgerCnt;
-            this.outstandings = outstandings;
-            this.syncObj = syncObj;
-            this.progressable = progressable;
+            outstanding.add(promise);
+            promise.whenComplete((metadata, ex) -> outstanding.remove(promise));
```

Comment:
this is not 100% correct. 

1) you are attaching a complete closure to `promise`, not to the promise after `thenCompose` above.

2) remove is mutating the structure. I am not sure how it would impact `outstanding.stream()` below.

so at this point I am not 100% confident about the refactor here. what I would suggest:

1) move `outstanding.remove` to the `whenComplete` above.  we don't need a separate whenComplete. that means you also move `outstanding.add` after constructing `promise`.

2) I would change line 144 to use a CountDownLatch rather than `CompletableFuture.allOf`.

Reply:
1. changed, though slightly different to what you suggested, as we need the write promise, it's the promise returned from thenCompose that is added to outstanding.

2. outstanding is a concurrenthashmap, wrapped in a set. the stream will be generated from an iterator on the ConcurrentHashMap, which is a view of the contents at or after the point of the iterator creation. These are safe from concurrentmodificationexception. At the point stream() is called, nothing new will be added to the set, though entries may be removed. This is fine, because remove will only remove a completed future, which is what we're waiting for anyhow. I'm not even sure how I could use CountDownLatch here, and we would end up having some messy rubbish to get any thrown error out.